translator/trace/jaeger/jaegerproto_to_protospan.go

```diff
@@ -0,0 +1,274 @@
+			Attributes:   sAttributes,
+			// TODO: StackTrace: OpenTracing defines a semantic key for "stack", should we attempt to its content to StackTrace?
+			TimeEvents: jProtoLogsToOCProtoTimeEvents(jspan.Logs),
+			Links:      jProtoReferencesToOCProtoLinks(jspan.References),
+			Status:     sStatus,
+		}
+
+		spans = append(spans, span)
+	}
+	return spans
+}
+
+func jProtoLogsToOCProtoTimeEvents(logs []model.Log) *tracepb.Span_TimeEvents {
+	if logs == nil {
+		return nil
+	}
+
+	timeEvents := make([]*tracepb.Span_TimeEvent, 0, len(logs))
+
+	for _, log := range logs {
+		description, _, _, attribs := jProtoTagsToAttributes(log.Fields)
+		var annotation *tracepb.Span_TimeEvent_Annotation
+		if attribs != nil {
+			annotation = &tracepb.Span_TimeEvent_Annotation{
+				Description: strToTruncatableString(description),
+				Attributes:  attribs,
+			}
+		}
+		timeEvent := &tracepb.Span_TimeEvent{
+			Time:  internal.TimeToTimestamp(log.Timestamp),
+			Value: &tracepb.Span_TimeEvent_Annotation_{Annotation: annotation},
+		}
+
+		timeEvents = append(timeEvents, timeEvent)
+	}
+
+	return &tracepb.Span_TimeEvents{TimeEvent: timeEvents}
+}
+
+func jProtoReferencesToOCProtoLinks(jrefs []model.SpanRef) *tracepb.Span_Links {
+	if jrefs == nil {
+		return nil
+	}
+
+	links := make([]*tracepb.Span_Link, 0, len(jrefs))
+
+	for _, jref := range jrefs {
+		var linkType tracepb.Span_Link_Type
+		if jref.RefType == model.SpanRefType_CHILD_OF {
+			// Wording on OC for Span_Link_PARENT_LINKED_SPAN: The linked span is a parent of the current span.
+			linkType = tracepb.Span_Link_PARENT_LINKED_SPAN
+		} else {
+			// TODO: SpanRefType_FOLLOWS_FROM doesn't map well to OC, so treat all other cases as unknown
+			linkType = tracepb.Span_Link_TYPE_UNSPECIFIED
+		}
+
+		link := &tracepb.Span_Link{
+			TraceId: tracetranslator.UInt64ToByteTraceID(jref.TraceID.High, jref.TraceID.Low),
+			SpanId:  tracetranslator.UInt64ToByteSpanID(uint64(jref.SpanID)),
+			Type:    linkType,
+		}
+		links = append(links, link)
+	}
+
+	return &tracepb.Span_Links{Link: links}
+}
+
+func jProtoTagsToAttributes(tags []model.KeyValue) (string, tracepb.Span_SpanKind, *tracepb.Status, *tracepb.Span_Attributes) {
+	if tags == nil {
+		return "", tracepb.Span_SPAN_KIND_UNSPECIFIED, nil, nil
+	}
+
+	var sKind tracepb.Span_SpanKind
+
+	var statusCodePtr *int32
+	var statusMessage string
+	var httpStatusCodePtr *int32
+	var httpStatusMessage string
+	var message string
+
+	sAttribs := make(map[string]*tracepb.AttributeValue)
+
+	for _, tag := range tags {
+		// Take the opportunity to get the "span.kind" per OpenTracing spec, however, keep it also on the attributes.
+		// TODO: (@pjanotti): Replace any OpenTracing literals by importing github.com/opentracing/opentracing-go/ext?
+		switch tag.Key {
+		case "span.kind":
+			switch tag.GetVStr() {
+			case "client":
+				sKind = tracepb.Span_CLIENT
+				continue // TODO(@jpkrohling): is it appropriate to return here?
```

Comment:
I used another Jaeger translator as base for this one (`jtagsToAttributes()`), and it does *not* continue to the next iteration there, but keeps processing the current iteration. The end result is that this here will remove the `span.kind` from the attributes and keep at the main model (`Kind` attribute), whereas the Thrift one will keep this data *both* inside the `Kind` attribute and as a tag.

I believe it's more appropriate to keep it only in the `Kind` attribute, but if this is wrong, let me know and I'll remove this `continue` from here.

Reply:
A recent change,  #189, ensures that it doesn't get duplicated in the final conversion if the export destination is Jaeger.

While in principle `Kind` can be converted back to the proper "tag" (attribute on OTel parlance) not all trace destinations follow the OpenTracing convention so preserving the original tag in this case seems reasonable (this only matters for deployments using different trace formats), while the final conversion from internal format to export format takes proper care of duplicates and so on.

Reply:
Got it, I'll remove the `continue` then and let it be handled down the chain.